When --path is used, the directory's .swiftlint.yml is ignored

When using SwiftLint 0.20.0, when specifying a path (that is not the current directory) with `--path`, that directory's `.swiftlint.yml` is ignored.

To reproduce it, just create the following 2 files in a subdirectory.

```yml
# subdir/.swiftlint.yml
disabled_rules:
  - todo
```

```swift
// subdir/MyCode.swift
// TODO: ABCDEFG
```

Specifying the path causes the directory's `.swiftlint.yml` to be ignored.
```
$ swiftlint lint --no-cache --quiet # Not path specified, no warning as expected
$ swiftlint lint --no-cache --quiet --path subdir # Specifying the path gets you an unexpected warning
.../subdir/MyCode.swift:2:4: warning: Todo Violation: TODOs should be avoided (ABCDEFG). (todo)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -28,6 +28,10 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#1630](https://github.com/realm/SwiftLint/issues/1630)
 
+* Use the directory's .swiftlint.yml when `--path` is used.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#1631](https://github.com/realm/SwiftLint/issues/1631)
+
 ## 0.20.0: Liquid Fabric Softener
 
 ##### Breaking

--- a/Source/SwiftLintFramework/Models/Configuration.swift
+++ b/Source/SwiftLintFramework/Models/Configuration.swift
@@ -155,7 +155,13 @@ public struct Configuration: Equatable {
 
     public init(path: String = Configuration.fileName, rootPath: String? = nil,
                 optional: Bool = true, quiet: Bool = false, enableAllRules: Bool = false, cachePath: String? = nil) {
-        let fullPath = path.bridge().absolutePathRepresentation()
+        let fullPath: String
+        if let rootPath = rootPath {
+            fullPath = path.bridge().absolutePathRepresentation(rootDirectory: rootPath)
+        } else {
+            fullPath = path.bridge().absolutePathRepresentation()
+        }
+
         let fail = { (msg: String) in
             queuedPrintError("\(fullPath):\(msg)")
             fatalError("Could not read configuration file at path '\(fullPath)'")
@@ -184,13 +190,6 @@ public struct Configuration: Equatable {
         }
         self.init(enableAllRules: enableAllRules, cachePath: cachePath)!
         setCached(atPath: fullPath)
-    }
-
-    public init(commandLinePath: String, rootPath: String? = nil, quiet: Bool = false, enableAllRules: Bool = false,
-                cachePath: String? = nil) {
-        self.init(path: commandLinePath, rootPath: rootPath?.absolutePathStandardized(),
-                  optional: !CommandLine.arguments.contains("--config"), quiet: quiet, enableAllRules: enableAllRules,
-                  cachePath: cachePath)
     }
 
     public func lintablePaths(inPath path: String, fileManager: LintableFileManager = FileManager.default) -> [String] {

--- a/Source/swiftlint/Commands/RulesCommand.swift
+++ b/Source/swiftlint/Commands/RulesCommand.swift
@@ -41,7 +41,7 @@ struct RulesCommand: CommandProtocol {
             return .success()
         }
 
-        let configuration = Configuration(commandLinePath: options.configurationFile)
+        let configuration = Configuration(options: options)
         let rules = ruleList(for: options, configuration: configuration)
 
         print(TextTable(ruleList: rules, configuration: configuration).render())
@@ -71,7 +71,7 @@ struct RulesCommand: CommandProtocol {
 
 struct RulesOptions: OptionsProtocol {
     fileprivate let ruleID: String?
-    fileprivate let configurationFile: String
+    let configurationFile: String
     fileprivate let filterEnabled: Bool
 
     static func create(_ configurationFile: String) -> (_ ruleID: String) -> (_ filterEnabled: Bool) -> RulesOptions {

--- a/Source/swiftlint/Extensions/Configuration+CommandLine.swift
+++ b/Source/swiftlint/Extensions/Configuration+CommandLine.swift
@@ -122,7 +122,9 @@ extension Configuration {
 
     init(options: LintOptions) {
         let cachePath = options.cachePath.isEmpty ? nil : options.cachePath
-        self.init(commandLinePath: options.configurationFile, rootPath: options.path, quiet: options.quiet,
+        let optional = !CommandLine.arguments.contains("--config")
+        self.init(path: options.configurationFile, rootPath: options.path.absolutePathStandardized(),
+                  optional: optional, quiet: options.quiet,
                   enableAllRules: options.enableAllRules, cachePath: cachePath)
     }
 
@@ -137,7 +139,15 @@ extension Configuration {
 
     init(options: AutoCorrectOptions) {
         let cachePath = options.cachePath.isEmpty ? nil : options.cachePath
-        self.init(commandLinePath: options.configurationFile, rootPath: options.path,
-                  quiet: options.quiet, cachePath: cachePath)
+        let optional = !CommandLine.arguments.contains("--config")
+        self.init(path: options.configurationFile, rootPath: options.path.absolutePathStandardized(),
+                  optional: optional, quiet: options.quiet, cachePath: cachePath)
+    }
+
+    // MARK: Rules command
+
+    init(options: RulesOptions) {
+        let optional = !CommandLine.arguments.contains("--config")
+        self.init(path: options.configurationFile, optional: optional)
     }
 }

--- a/Tests/LinuxMain.swift
+++ b/Tests/LinuxMain.swift
@@ -67,6 +67,7 @@ extension ConfigurationTests {
         ("testLevel1", testLevel1),
         ("testLevel2", testLevel2),
         ("testLevel3", testLevel3),
+        ("testNestedConfigurationWithCustomRootPath", testNestedConfigurationWithCustomRootPath),
         ("testCustomConfiguration", testCustomConfiguration),
         ("testConfiguresCorrectlyFromDict", testConfiguresCorrectlyFromDict),
         ("testConfigureFallsBackCorrectly", testConfigureFallsBackCorrectly),

--- a/Tests/SwiftLintFrameworkTests/ConfigurationTests.swift
+++ b/Tests/SwiftLintFrameworkTests/ConfigurationTests.swift
@@ -154,13 +154,17 @@ class ConfigurationTests: XCTestCase {
     }
 
     private var projectMockConfig0CustomPath: Configuration {
-        var configuration = Configuration(path: projectMockYAML0CustomPath, optional: false, quiet: true)
-        configuration.rootPath = projectMockPathLevel0
-        return configuration
+        return Configuration(path: projectMockYAML0CustomPath, rootPath: projectMockPathLevel0,
+                             optional: false, quiet: true)
     }
 
     fileprivate var projectMockConfig2: Configuration {
         return Configuration(path: projectMockYAML2, optional: false, quiet: true)
+    }
+
+    fileprivate var projectMockConfig3: Configuration {
+        return Configuration(path: Configuration.fileName, rootPath: projectMockPathLevel3,
+                             optional: false, quiet: true)
     }
 
     func testIsEqualTo() {
@@ -194,7 +198,12 @@ class ConfigurationTests: XCTestCase {
 
     func testLevel3() {
         XCTAssertEqual(projectMockConfig0.configuration(for: File(path: projectMockSwift3)!),
-                       projectMockConfig0.merge(with: projectMockConfig2))
+                       projectMockConfig0.merge(with: projectMockConfig3))
+    }
+
+    func testNestedConfigurationWithCustomRootPath() {
+        XCTAssertEqual(projectMockConfig3,
+                       projectMockConfig0.merge(with: projectMockConfig3))
     }
 
     // MARK: - Testing Custom Configuration File

--- a/Tests/SwiftLintFrameworkTests/Resources/ProjectMock/Level1/Level2/Level3/.swiftlint.yml
+++ b/Tests/SwiftLintFrameworkTests/Resources/ProjectMock/Level1/Level2/Level3/.swiftlint.yml
@@ -1,0 +1,9 @@
+disabled_rules:
+  - force_try
+  - todo
+included:
+  - "everything on level 3"
+excluded:
+  - "the place where i committed many coding sins"
+line_length: 10000000000
+reporter: "json"

